feat: lessons about crawling and scraping product detail pages in Python

honzajavorek opened the pull request:
### Done

This PR introduces two new lessons to the Python course, including real-world exercises. These two lessons conclude the base of the course, as at the end of the course the reader should be able to build their own scraper. The final exercises focus on that fact and test the reader's ability of independent building. The PR also includes some edits to the previous parts of the course (code examples, exercises).

### Next

Before the course is done, there should be two more lessons: One about building the very same scraper using a framework (Crawlee), and one about deploying the scraper to a platform (Apify). Then I should return back to the beginning and complete the three initial lessons about DevTools.

honzajavorek commented on sources/academy/webscraping/scraping_basics_python/04_downloading_html.md:
```diff
@@ -142,10 +142,10 @@ Letting our program visibly crash on error is enough for our purposes. Now, let'
 ### Scrape Amazon
 
-Download HTML of a product listing page, but this time from a real world e-commerce website. For example this page with Amazon search results:
+Download HTML of a product listing page, but this time from a real world e-commerce website. For example this page with AliExpress search results:
```
I decided to do this because this is like... the very first exercise, and I don't want the reader to end up with an error. I don't want to direct them somewhere else when they're touching the internet for the first time with their tiny program. So hoping for better results with Ali, which is a real world e-commerce website as well, but doesn't seem to give so much damn about scraping.

honzajavorek commented on sources/academy/webscraping/scraping_basics_python/09_getting_links.md:
```diff
@@ -199,8 +199,12 @@ def export_json(file, data):
 listing_url = "https://warehouse-theme-metal.myshopify.com/collections/sales"
-soup = download(listing_url)
-data = [parse_product(product) for product in soup.select(".product-item")]
+listing_soup = download(listing_url)
```
I made this change because this way the code is better suited for the newly added lessons.

mnmkng commented on sources/academy/webscraping/scraping_basics_python/11_scraping_variants.md:
````diff
@@ -0,0 +1,418 @@
+]
+```
+
+Some products where we're missing the actual price should now have several variants:
````
I actually didn't understand why the price was missing and thought it was a bug until I read to the section where you add the prices.

```suggestion
Some products where we're missing the actual price will now have several variants. We will scrape the price later for those products.
```

honzajavorek replied:
That's a good catch as well. I'll re-read the lesson and try to make this part more comprehensible.

honzajavorek replied:
Addressed in [b82b673](https://github.com/apify/apify-docs/pull/1244/commits/b82b673a31d8a7769e5b52c99a94a50b4480029a), thanks!

mnmkng reviewed (changes requested):
Thanks Honza, great job as always.

honzajavorek commented on sources/academy/webscraping/scraping_basics_python/04_downloading_html.md:
```diff
@@ -142,10 +142,10 @@ Letting our program visibly crash on error is enough for our purposes. Now, let'
 
 ### Scrape Amazon
```
Good catch! Thanks

honzajavorek commented:
I addressed all comments from @mnmkng. I also added a gif as I thought it will make one part clearer. I'm not a seasoned gif creator, so if someone can make a more beautiful one, go for it. Attaching the source recording as a file.

https://github.com/user-attachments/assets/551450ea-7f27-460b-b82e-0a64c0c39687

This is the gif:

![variants-js](https://github.com/user-attachments/assets/1bc3b980-c837-4adf-b5da-13cc871eec74)

I made it like this:

```
ffmpeg -i Screen\ Recording\ 2024-10-18\ at\ 9.42.13.mov -pix_fmt rgb24 -r 10 -f gif - | gifsicle --optimize=2 --delay=10 > variants-js.gif
```

Unfortunately it's not possible to show how the price changes dynamically, as the JS touches the parent elements and in both Chrome and Firefox devtools this causes the elements to collapse. At least it's visible that _something_ changes there, and how the text of the stock availability changes.

mnmkng commented:
@vdusek can you please check the Python so that we can merge? Thanks!

vdusek commented on sources/academy/webscraping/scraping_basics_python/10_crawling.md:
````diff
@@ -1,15 +1,305 @@
+
+In the `data` loop we're already going through all the products. Let's expand it to include downloading the product detail page, parsing it, extracting the vendor's name, and adding it as a new key in the item's dictionary:
+
+```py
+...
+
+listing_url = "https://warehouse-theme-metal.myshopify.com/collections/sales"
+listing_soup = download(listing_url)
+
+data = []
+for product in listing_soup.select(".product-item"):
+    item = parse_product(product, listing_url)
+    # highlight-next-line
+    product_soup = download(item["url"])
+    # highlight-next-line
+    item["vendor"] = product_soup.select_one(".product-meta__vendor").text.strip()
+    data.append(item)
+
+...
+```
````
It runs surprisingly long:

```text
$ time python review_pr_1244.py 

real    0m6,548s
user    0m1,917s
sys     0m0,070s
```

for scraping ~20 items. But I believe it's the website, right?

mnmkng replied:
The website is actually quite fast. At least for me 😅

mnmkng replied:
What about merging this and creating an issue for the remaining problems so that it can go on-line and start getting indexed, used and so on?

honzajavorek replied:
Once it's in the git history, the repo gets bloated forever. I'm for optimizing the images first, unless me pushing to this branch didn't already do the damage.

honzajavorek replied:
I'd even do it off my stint, but unfortunately my family situation is just getting worse, so I barely work on anything. OTOH anyone who's able to run CLI commands (e.g. [this](https://github.com/343dev/optimizt)) and push to branches should be able to do it.

honzajavorek replied:
I didn't measure the speed of the script, but yes, it's slow and it's kinda expected, it's just a loop with requests. It's gonna be one of the reasons why to employ Crawlee in the next lesson.

I see the size of the images as a bigger problem. Once it's in the git history, the repo gets bloated forever. I'm for optimizing the images first, unless me pushing to this branch didn't already do the damage.

I'd even do it off my stint, but unfortunately my family situation is just getting worse, so I barely work on anything. OTOH anyone who's able to run CLI commands (e.g. [this](https://github.com/343dev/optimizt)) and push to branches should be able to do it.

honzajavorek replied:
```
+243 KB sources/academy/webscraping/scraping_basics_python/images/pdp.png
+494 KB sources/academy/webscraping/scraping_basics_python/images/variants.png
```

vdusek reviewed (changes requested):
The code looks good and works well. I just have one suggestion regarding the images...

Currently, the images are quite large (~1.2 MB each). Could we apply some compression? I would say using `jpeg` would be enough. To avoid unnecessary bloating of the Git history.

honzajavorek commented:
I applied compression to the PNG files added in this PR, results being:
```
+243 KB sources/academy/webscraping/scraping_basics_python/images/pdp.png
+494 KB sources/academy/webscraping/scraping_basics_python/images/variants.png
```
I rebased the branch so that it contains only these new files, I hope that's gonna be sufficient. @vdusek @mnmkng if you see no further problems with the PR, bless this with green and I think we can merge.

TC-MO commented on sources/academy/webscraping/scraping_basics_python/06_locating_elements.md:
```diff
@@ -122,6 +122,14 @@ for product in soup.select(".product-item"):
 
 This program does the same as the one we already had, but its code is more concise.
 
+:::note Fragile code
```
All valid, then note makes perfect sense

vdusek reviewed (approved):
LGTM
